3party/opening_hours/rules_evaluation.cpp

```diff
@@ -385,6 +385,150 @@ bool IsActive(RuleSequence const & rule, time_t const timestamp)
   return res.first && res.second;
 }
 
+time_t GetNextTimeOpen(TRuleSequences const & rules, time_t const dateTime)
+{
+  const time_t TIME_T_MAX = std::numeric_limits<time_t>::max();
+  time_t dateTimeResult = TIME_T_MAX;
+  time_t dateTimeToCheck;
+
+  // Check in the next 7 days
+  for (int i = 0; i < 7; i++)
+  {
+    for (auto it = rules.rbegin(); it != rules.rend(); ++it)
+    {
+      auto const & times = it->GetTimes();
+
+      // If the rule has no times specified, check at 00:00
+      if (times.size() == 0)
+      {
+        tm tm = MakeTimetuple(dateTime);
+        tm.tm_hour = 0;
+        tm.tm_min = 0;
+        dateTimeToCheck = mktime(&tm) + i * (24 * 60 * 60);
+
+        if (dateTimeToCheck < dateTime || dateTimeToCheck > dateTimeResult)
+          continue;
+
+        if (GetState(rules, dateTimeToCheck) == RuleState::Open)
+          dateTimeResult = dateTimeToCheck;
+      }
+
+      // If the rule is of "Closed" type, check ending times
+      if (it->GetModifier() == RuleSequence::Modifier::Closed)
+      {
+        // Check the ending time of each rule
+        for (auto const &time : times)
+        {
+          tm tm = MakeTimetuple(dateTime);
+          tm.tm_hour = time.GetEnd().GetHourMinutes().GetHoursCount();
+          tm.tm_min = time.GetEnd().GetHourMinutes().GetMinutesCount();
+          dateTimeToCheck = mktime(&tm) + i * (24 * 60 * 60) + 60;  // 1 minute offset
+          
+          if (dateTimeToCheck < dateTime || dateTimeToCheck > dateTimeResult)
+            continue;
+
+          if (GetState(rules, dateTimeToCheck) == RuleState::Open)
+            dateTimeResult = dateTimeToCheck;
+        }
+      }
+      else
+      {
+        // Check the starting time of each rule
+        for (auto const &time : times)
+        {
+          tm tm = MakeTimetuple(dateTime);
+          tm.tm_hour = time.GetStart().GetHourMinutes().GetHoursCount();
+          tm.tm_min = time.GetStart().GetHourMinutes().GetMinutesCount();
+          dateTimeToCheck = mktime(&tm) + i * (24 * 60 * 60);
+          
+          if (dateTimeToCheck < dateTime || dateTimeToCheck > dateTimeResult)
+            continue;
+
+          if (GetState(rules, dateTimeToCheck) == RuleState::Open)
+            dateTimeResult = dateTimeToCheck;
+        }
+      }
+    }
+
+    if (dateTimeResult < TIME_T_MAX)
+      return dateTimeResult;
+  }
+
+  return TIME_T_MAX;
+}
+
+time_t GetNextTimeClosed(TRuleSequences const & rules, time_t const dateTime)
```

Comment:
This function looks very similar to GetNextTimeOpen, is it possible to avoid a copy-paste?

Reply:
I merged both functions into a single GetNextTimeState function